plugins/obs-qsv11/obs-qsv265.c

```diff
@@ -0,0 +1,1118 @@
+#define info(format, ...) do_log(LOG_INFO, format, ##__VA_ARGS__)
+#define debug(format, ...) do_log(LOG_DEBUG, format, ##__VA_ARGS__)
+
+static uint8_t *ff_find_startcode_internal(const uint8_t *p, const uint8_t *end)
+{
+	const uint8_t *a = p + 4 - ((intptr_t)p & 3);
+
+	for (end -= 3; p < a && p < end; p++) {
+		if (p[0] == 0 && p[1] == 0 && p[2] == 1)
+			return (uint8_t *)p;
+	}
+
+	for (end -= 3; p < end; p += 4) {
+		uint32_t x = *(const uint32_t *)p;
+
+		if ((x - 0x01010101) & (~x) & 0x80808080) {
+			if (p[1] == 0) {
+				if (p[0] == 0 && p[2] == 1)
+					return (uint8_t *)p;
+				if (p[2] == 0 && p[3] == 1)
+					return (uint8_t *)(p + 1);
+			}
+
+			if (p[3] == 0) {
+				if (p[2] == 0 && p[4] == 1)
+					return (uint8_t *)(p + 2);
+				if (p[4] == 0 && p[5] == 1)
+					return (uint8_t *)(p + 3);
+			}
+		}
+	}
+
+	for (end += 3; p < end; p++) {
+		if (p[0] == 0 && p[1] == 0 && p[2] == 1)
+			return (uint8_t *)p;
+	}
+
+	return (uint8_t *)(end + 3);
+}
+
+static uint8_t *obs_find_annexB_startcode(const uint8_t *p, const uint8_t *end)
+{
+	uint8_t *out = ff_find_startcode_internal(p, end);
+	if (p < out && out < end && !out[-1])
+		out--;
+	return out;
+}
+
+/* ------------------------------------------------------------------------- */
+
+struct obs_qsv {
+	obs_encoder_t *encoder;
+
+	qsv_param_t params;
+	qsv_t *context;
+
+	DARRAY(uint8_t) packet_data;
+
+	uint8_t *extra_data;
+	uint8_t *sei;
+
+	size_t extra_data_size;
+	size_t sei_size;
+
+	os_performance_token_t *performance_token;
+};
+
+/* ------------------------------------------------------------------------- */
+
+static CRITICAL_SECTION g_QsvCs;
+static unsigned short g_verMajor;
+static unsigned short g_verMinor;
+static int64_t g_pts2dtsShift;
+static int64_t g_prevDts;
+static bool g_bFirst;
+
+static const char *obs_qsv_hevc_getname(void *type_data)
+{
+	UNUSED_PARAMETER(type_data);
+	return "QuickSync HEVC Encoder";
+}
+
+static void obs_qsv_hevc_stop(void *data);
+
+static void clear_data(struct obs_qsv *obsqsv)
+{
+	if (obsqsv->context) {
+		EnterCriticalSection(&g_QsvCs);
+		qsv_hevc_encoder_close(obsqsv->context);
+		obsqsv->context = NULL;
+		LeaveCriticalSection(&g_QsvCs);
+
+		bfree(obsqsv->sei);
+		obsqsv->sei = NULL;
+
+		bfree(obsqsv->extra_data);
+		obsqsv->extra_data = NULL;
+	}
+}
+
+static void obs_qsv_hevc_destroy(void *data)
+{
+	struct obs_qsv *obsqsv = (struct obs_qsv *)data;
+
+	if (obsqsv) {
+		os_end_high_performance(obsqsv->performance_token);
+		clear_data(obsqsv);
+		da_free(obsqsv->packet_data);
+		bfree(obsqsv);
+	}
+}
+
+static void obs_qsv_hevc_defaults(obs_data_t *settings)
+{
+	obs_data_set_default_string(settings, "target_usage", "veryfast");
+	obs_data_set_default_int(settings, "bitrate", 2500);
+	obs_data_set_default_int(settings, "max_bitrate", 30000);
+	obs_data_set_default_string(settings, "profile", "Main 10");
+	obs_data_set_default_string(settings, "rate_control", "VBR");
```

Comment:
Default for H.264 is "CBR". Why change to "VBR" for HEVC?

Reply:
I tried CBR and VBR for HEVC/HDR (using YouTube as streaming platform) and got better results for VBR (in sense of CPU usage/network utilization) for 4K/60.
For sure, this is not absolutely correct answer because it depends of other options and is not . But my current point of view is VBR.

So, leave it as is or change to?

Reply:
This is Twitch's reason for preferring CBR to VBR in a streaming scenario, and it probably applies to YouTube streaming too:
> You may have noticed the option to select between VBR (variable bitrate) and CBR (constant bitrate). Twitch suggests that all Broadcasters use CBR for several reasons, all of which relate to the final quality of service (QoS) that your viewers will experience.

> The main problem with VBR is that during lulls in the action: paused games, hero selection screens, etc, VBR streams produce a significantly lower bitrate. This can cause issues (stream buffering, dropped frames) with many end-user connections when the bitrate spikes back up during the action (team fights, high motion scenes).

> VBR also leads to issues when sending data to the Twitch network over many ISPs. When connecting to Twitch, the route your ISP takes may not be stable enough to handle sudden increases in bandwidth, leading to “broadcast starvation.”

> Broadcast starvation (instability) is when video data does not arrive in a timely fashion or is completely missing. This happens when there are network problems between you and Twitch, when you attempt to stream at a bitrate that is too high for your network/ISP, or have problems with your router causing frames to be dropped before entering the Twitch network. This leads to lag for everyone (including people watching on your lower resolutions), buffering, and stream disconnections.

Reply:
Reverting to default for now.